File: docs/article/source/ruby/apps/cargo-tracker/packs/booking/app/domain/booking/domain/cargo_itinerary.rb
# frozen_string_literal: true

module Booking
  module Domain
    # 旅程（CargoItinerary）。1 区間以上の Leg 列からなり、隣接脚は連結（Leg[n].unload == Leg[n+1].load）する。
    # 出発地・目的地・到着予定時刻を導出する不変オブジェクト。
    CargoItinerary = Data.define(:legs) do
      def initialize(legs:)
        raise ArgumentError, "旅程には 1 区間以上の脚が必要です" if legs.nil? || legs.empty?

        legs.each_cons(2) do |current, following|
          if current.unload_location != following.load_location
            raise ArgumentError,
                  "脚が連結しません: #{current.unload_location} → #{following.load_location}"
          end
        end

        # 到着期限判定（satisfied_by?）に用いるため最終脚の荷揚時刻は必須とする。
        raise ArgumentError, "最終脚の荷揚時刻は必須です" if legs.last.unload_time.nil?

        super
      end

      def origin = legs.first.load_location
      def destination = legs.last.unload_location
      def expected_arrival_time = legs.last.unload_time
    end
  end
end

File: docs/article/source/java-4/apps/cargo-tracker/backend/bookingms/src/main/java/com/example/cargotracker/booking/domain/model/valueobjects/CargoItinerary.java
package com.example.cargotracker.booking.domain.model.valueobjects;

import com.example.cargotracker.shared.domain.error.BusinessRuleViolation;
import com.example.cargotracker.shared.domain.location.Location;
import java.time.Instant;
import java.util.List;

/**
 * 確定した旅程（US09。domain-model.md「Cargo 集約の不変条件」4）。
 *
 * <p><b>連結と時刻の昇順はここで守る。</b> 経路探索も同じ検査をするが、探索が作った
 * ものだけが旅程になるとは限らない（誤配の再設計・API を直接叩く経路）。
 * 「候補は探索が作ったのだから正しい」としない。</p>
 */
public record CargoItinerary(List<Leg> legs) {

    public CargoItinerary {
        if (legs == null || legs.isEmpty()) {
            throw new BusinessRuleViolation("旅程は 1 区間以上が必要です");
        }
        legs = List.copyOf(legs);
        for (int i = 1; i < legs.size(); i++) {
            Leg previous = legs.get(i - 1);
            Leg current = legs.get(i);
            if (!previous.unload().equals(current.load())) {
                throw new BusinessRuleViolation("区間が連結していません: "
                        + previous.unload().unLocode().value() + " → "
                        + current.load().unLocode().value());
            }
            if (current.loadTime().isBefore(previous.unloadTime())) {
                throw new BusinessRuleViolation("前の区間の到着より前に出発する区間があります: "
                        + previous.unloadTime() + " → " + current.loadTime());
            }
        }
    }

    /** 最初の積地。 */
    public Location origin() {
        return legs.get(0).load();
    }

    /** 最後の揚地。 */
    public Location destination() {
        return legs.get(legs.size() - 1).unload();
    }

    /** 最後の到着日時。期限を満たすかの判断に使う。 */
    public Instant finalArrival() {
        return legs.get(legs.size() - 1).unloadTime();
    }
}
